### ISSUE TITLE
Issues with security upgrade - Random cache issues, and auto-release

### ISSUE THREAD

**Issue by asuweb** [mailwatch/1.2.0#662]
The new updates to improve security seem to be causing some issues for my users.

It's seems that caching can prevent pages from loading and they return "No! Bad dog no treat for you!"  - it's not the most helpful of phrases or expressions - and has resulted in a number of support tickets this morning.

Also, the auto-release function no longer works with these upgrades resulting in the same phrase.

I haven't time right now to debug them, but if anyone else happens to have a moment, I'd appreciate it.

**Comment by shawniverson** [comment]
When you get a chance....does clearing the browser cache on an affected
machine yield any results?

I'm going to enable auto-release and see if I can reproduce.

**Comment by asuweb** [comment]
Clearing the cache or using an incognito window solves the cache issue, but doesn't fix auto-release

**Comment by shawniverson** [comment]
Yeah, found a couple of issues....working through them, will prepare PR
shortly.

**Comment by asuweb** [comment]
Thanks @shawniverson

**Comment by shawniverson** [comment]
Now for the bad dog message....yeah, we probably should change that, in fact a message to clear your cache would be helpful.   Something like "Error accessing MailWatch, please clear your browser cache and try again" or similar.

**Comment by endelwar** [comment]
> Now for the bad dog message....yeah, we probably should change that

eheh, yes, we should add 2 different error messages: one for token error and another for input validator error, so that we can offer a better support to the ones that encounter these problems

**Comment by stefaweb** [comment]
Last develop version (#676).

`Error: unable to validate security token` when I try to login.

Have to quit browser (Safari OS X) and try again. Works after.

**Comment by shawniverson** [comment]
@stefaweb Did it happen a single time and was following application of the latest develop to your server?

**Comment by stefaweb** [comment]
Yes, one time and after MW upgrade on my dev server.

**Comment by shawniverson** [comment]
Sounds caching related.  It doesn't look like login.php/checklogin.php has no-cache header set, so that might help if we add it.

**Comment by stefaweb** [comment]
Still have `Error: unable to validate security token` per moment.

The process.

After a moment, MailWatch auto disconnect and go back to login.php. May be when I let MW not on status.page.

I click on `Login` button (field autofiled), go to a new white page and got `Error: unable to validate security token`.

If I reload the page, I go back to login.php and got this screen.

![capture d ecran 2017-03-24 a 17 03 25](https://cloud.githubusercontent.com/assets/12907102/24302753/02ef6b36-10b4-11e7-8790-cace87e1276f.png)

If I click on `Login` button again, I got the `Bad user name` page again.

If I click a second time, I go inside the GUI on status.page.

Good luke!

**Comment by shawniverson** [comment]
@stafaweb, are you letting the page sit until the session expires to get
this pattern?  (want to try same steps)

**Comment by shawniverson** [comment]
Sorry about misspell on my cell :)

**Comment by asuweb** [comment]
@shawniverson - I have the same issue, and yes, it's due to the session expiry.  

I leave the status page open anytime I'm at my desk (often for hours) which previously wasn't an issue, but the session times out once an hour now, which then causes this loop (only sometimes though) .

**Comment by shawniverson** [comment]
@endelwar,  I'm pretty sure you don't want sessions hanging around forever,
but I wonder if this can somehow be improved or even adjustable?

I'm going to make the session expire quickly on my end to rapid fire the
loop issue and work on it.

**Comment by asuweb** [comment]
I haven't been through the code new code fully yet, but how are we currently setting sessions expiry?

**Comment by shawniverson** [comment]
@asuweb,

In functions.php, it appears...3600 seconds here.

session_set_cookie_params(60 * 60, $params['path'], $params['domain'],
$session_cookie_secure, true);

**Comment by asuweb** [comment]
Would it perhaps be better to set `session.cookie.lifetime` to `0` which would destroy the session on browser close, and then control session expiry in `checklogin` and give sys-admins the option in `conf.php` to either set a fixed expiry time (like it is now of 3600s) or a activity based expiry like if no activity in x minutes, destroy session?

**Comment by shawniverson** [comment]
I think that makes a lot more sense, at least to me.

**Comment by asuweb** [comment]
Are you doing it or will I do it?

**Comment by endelwar** [comment]
session lifetime should be as short as possible, OWASP suggest 2-5 minutes for critical app, and 15 minutes for low risk app (https://www.owasp.org/index.php/Session_Management_Cheat_Sheet#Session_Expiration): we are at 60 minutes on an app that has full access to (private) email content.

We can find a solution on status.php page that _on automatic refresh_ regenerate session id

**Comment by stefaweb** [comment]
Status.php should work 24/24 without disconnect.
I've it on display in the NOC. 😎

**Comment by stefaweb** [comment]
And mailq.php

**Comment by asuweb** [comment]
@endelwar - I agree with you for in activity, but if left on the status page, I think we should keep the session alive (but have the option for sys-admins to enforce a fixed session if they want.  Could also be based on user type

**Comment by asuweb** [comment]
@stefaweb - mailq doesn't automatically refresh with all setups (postfix for example)

**Comment by asuweb** [comment]
@endelwar - also, whilst I agree with OWASP recommendation on expiry - I don't know of many webmail clients or web control panels which have such short sessions - they'd be dysfunctional if they did.

**Comment by endelwar** [comment]
a quick fix that jump in my mind is that status.php refresh to index.php (where token and session is regenerated) and that automatically forwarded to status.php again

**Comment by stefaweb** [comment]
I've added auto refresh to mailq.php

**Comment by asuweb** [comment]
I hadn't spotted that

**Comment by asuweb** [comment]
@stefaweb - did you do it also for `postfixmailq.php`?

**Comment by stefaweb** [comment]
> did you do it also for postfixmailq.php?

No, I don't.

I just added this in mail.php.

html_start(__('mqviewer24'), STATUS_REFRESH, false, false);

**Comment by asuweb** [comment]
@endelwar - that would be a quick fix, but not a permanent solution.  

I propose the following, in order of priority:

1. Reset session expiry on user activity

1. Reset session expiry on auto_refreshed pages

1. Optional configuration for fine grained control over cache expiry, auto_refresh keep alive, and admin control over which users/groups can keep sessions alive and those that can't - for example, it would be easy enough to implement a solution such as @stefaweb where a specific user can bypass session expiry (for the case where it's displays in a NOC or the likes).  

This way, we (as in MailWatch / EFa) would be providing a secure system, in line with guidance, but not restricting usability, by giving certain controls to the sysadmin to adapt the setup to their needs / requirements.

I'm happy to implement the above if wished, but I'll await further input before doing so

**Comment by shawniverson** [comment]
@stefaweb @asuweb Reproducing loop issue here.  I'm working on this part.

**Comment by shawniverson** [comment]
Problem found for looping.....silly little error in token check.

**Comment by shawniverson** [comment]
I also see bad timing occurring from expiration, which happens right away.  So.... a possible band-aid....right after functions.php, override the cookie expiration, and regenerate the session id....

```
$session_cookie_secure = false;
if (SSL_ONLY === true) {
    ini_set('session.cookie_secure', 1);
    $session_cookie_secure = true;
}
$params = session_get_cookie_params();
session_set_cookie_params(0, $params['path'], $params['domain'], $params['secure'], $params['httponly']);
session_set_cookie_params(0, $params['path'], $params['domain'], $session_cookie_secure, true);
unset($session_cookie_secure);

session_start();
disableBrowserCache();
session_regenerate_id(true);
```

However, this won't rectify a case where someone leaves login.php open for an extended period.  If session goes away for any reason, well, then no token.

**Comment by shawniverson** [comment]
Using the above code now for a few days and not having issues with login anymore.

**Comment by stefaweb** [comment]
Great news. You solved looping AND session expiration?

**Comment by shawniverson** [comment]
Just the looping.  Session expiration still will need addressed so that you don't get logged off every hour.

**Comment by endelwar** [comment]
> Using the above code now for a few days and not having issues with login anymore.

Mind opening a PR to test?

**Comment by endelwar** [comment]
> Reset session expiry on user activity

absolutely

> Reset session expiry on auto_refreshed pages

This too must be done
 
> Optional configuration for fine grained control over cache expiry, auto_refresh keep alive, and admin control over which users/groups can keep sessions alive and those that can't - for example, it would be easy enough to implement a solution such as @stefaweb where a specific user can bypass session expiry (for the case where it's displays in a NOC or the likes).

I see different way to achieve this, which way would you like to follow? (maybe open another issue to discuss this improvement)

**Comment by asuweb** [comment]
@endelwar - I'll work on the first 2 of those (unless anyone else already is?)  

I'll open a separate issue for discussion on the third

**Comment by shawniverson** [comment]
@asuweb Haven't done anything to address expiration issue so far.

**Comment by stefaweb** [comment]
Looping is gone for me.

**Comment by stefaweb** [comment]
Bad news.

I again got "Error: unable to validate security token" clicking on login.

MW was on status.php.
MW disconnect itself and go to loging page.
2 or 3 hours later, I clicked on the login button and got the infamous message.

**Comment by shawniverson** [comment]
@stefaweb This can still happen if the session goes away over an extended period.  The message in the case should really say "Login Session Expired, please login again."

**Comment by stefaweb** [comment]
I tried again. The problem is repetitive.

You login.
You stay in status.
When the session expire, MW quit status page and go to login page itself.
You wait.
You click on the login button (login/password autofilled by browser).
You go to the white page instead to be logged to the status page.

**Comment by asuweb** [comment]
I assume it's happening because the token isn't valid. Do we need to set the token on the login page or can that be set once the user has been validated?

**Comment by stefaweb** [comment]
Oops, it was difficult to login after that.

I was on page checklogin.pp, not login.php.

I removed chckloging.from php manually and hit enter. No luck, go back to the message.

I tried to reload page. Same issue.

I add to quit browser and go to the web site again.

**Comment by stefaweb** [comment]
I will see.

If you are on status/queue, no disconnect when session end. status page still continue to show mail and refresh.
If your click on something on the status page after session end (detail or menus), the system go to login page with a message  "Login session expired, please login again."
If you are on another page when the session expire, the system go to the login page automatically with a message  "Login session expired, please login again.".

**Comment by asuweb** [comment]
If the session has expired it should be redirecting to the login page on the next refresh in that case.

**Comment by shawniverson** [comment]
login.php should be doing the following in latest develop with the above fix.  Of course, checklogin.php cannot be accessed directly anymore.

1) Set session to not expire
2) Set a new session id
3) Generate a token unconditionally using the new session id (pre login)
4) Use this token in the post form
5) Redirect to checklogin.php and compare the post form token with session token
6) Generate a new (post login) session and token upon successfull login

@stefaweb It is almost as if your system is skipping the first 3 steps?  Another browser cache issue perhaps?  I am still having trouble reproducing.

**Comment by stefaweb** [comment]
I'll try this morning again. I uploaded last develop.

**Comment by stefaweb** [comment]
The problem is still here.

**Comment by shawniverson** [comment]
@stefaweb 

Is the problem still present without the autofill on?  Which browser/os are you on so I can trace this issue?

**Comment by stefaweb** [comment]
Mac OSX 10.12.4
Safari 10.1

I start a new test without autofill.

And after another one with Chrome.

**Comment by shawniverson** [comment]
Ok, I will do the same.  I have my Mac at work and will be heading in shortly.

**Comment by stefaweb** [comment]
No change if autofill OFF.

I'm trying with Chrome now.

**Comment by shawniverson** [comment]
Running a test with Safari, right now, so far no issues with or without autofill :?   However, it is not 10.1, so I will update now and try again.

Just to make sure I wasn't going crazy I did a diff of my php and the latest develop to make sure I didn't miss anything.

**Comment by stefaweb** [comment]
I'm late with the test. I was on others problems (difficult VPN setup..).

I restarted from scratch. I downloaded a fresh copy of develop using a new git directory on the dev server.

How to reduce session time in MW?

To reduce test time.

**Comment by shawniverson** [comment]
In functions.php, on line #101:

```
session_set_cookie_params(60 * 60, $params['path'], $params['domain'], $session_cookie_secure, true);
```

Change 60*60 (3600 seconds) to something else, like just 60 sec.

**Comment by stefaweb** [comment]
thanks.

**Comment by stefaweb** [comment]
Crazy story. It works with 120 seconds!!!

I switch back session to 1 hour to see.

**Comment by shawniverson** [comment]
Oh really?  Let me do that too

**Comment by stefaweb** [comment]
But I've updated with a fresh code. I should have done it without touching the code.

**Comment by asuweb** [comment]
There's another bug here - If we set `SSL_ONLY` in the config, and access the site via `http` the login page loads without ssl, and the token fails on login.

This may be a separate bug, as I would assume (I haven't yet looked) that requests to `http` when `SSL_ONLY` is set to `true` should be automatically redirected to `https` but that's not currently happening.

**Comment by stefaweb** [comment]
It works. Both on Safari and Chrome. It well go back to login.php instead of checkloging.php. And when I log, I no longer have the infamous message.

I cross my finger but seems gone with a fresh copy of develop.

Sorry for all this. A piece of code had to be patched badly in my dev copy.


For SSL, I do the rewrite in the vhost. Never tested SSL_ONLY.

**Comment by asuweb** [comment]
SSL issue is fixed in PR #710

**Comment by stefaweb** [comment]
`Error: unable to validate security token` is back!

If we reconnect to MW only after a long period, we got the message.

When I leaved my desk, MW was on status page. When I go back 2 hours later, MW was on the login page.

I clicked on login and got the message.

I launched a new test.

**Comment by shawniverson** [comment]
Yeah, that makes sense.  It isn't looping anymore, though, right?  Just one time and only after several hours.

Next step is some tweaking, but want to make sure you aren't getting stuck anymore.

**Comment by asuweb** [comment]
It works fine if it just logs you out and then you login with an hour of being logged out - because the new session generated on the login page is still valid.

This would be solved by managing session expiry internally rather than relying on cookies. There's no advantage to setting a session expiry on the login page, as there is no session to protect.  Could we perhaps generate the token on login submission rather than on page load?

**Comment by shawniverson** [comment]
@asuweb  The login page is set to not expire except on browser close (it is set to 0).  However, there are things besides closing the browser that cause this to happen.  The only way I see this going away completely is to dump the token/session logic during pre-login.  It is currently needed on page load because it is a hidden element of the login form generated during page load and then verified in checklogin.php upon submission.

**Comment by shawniverson** [comment]
Besides, it might make much better sense to instead look at a login attempt threshold and ditch this pre-login stuff to prevent brute force attempts on MailWatch login.  The token is only protecting against direct access to checklogin.php, so an attacker is just going to focus on login.php instead to do any brute forcing.

**Comment by asuweb** [comment]
@shawniverson - I hadn't checked that, but yes, I see you are right, it is set to 0 on the login page.

So if that's the case, why is the token mismatch happening?  The token is stored in both the post submission (hidden field in the login form) and the session variable - so they should match

**Comment by asuweb** [comment]
Brute force protection is far more important in my opinion.

**Comment by shawniverson** [comment]
@asuweb If anything similar to a browser close happens.....perhaps a sleep or standby.  My guess, haven't put it to the test yet.  But yes, I do see the session go away occasionally after extended periods.

**Comment by shawniverson** [comment]
If @endelwar is agreeable to the change, I propose dumping the pre-login token logic in favor of brute force prevention to protect the pre-login phase.

**Comment by asuweb** [comment]
@shawniverson - yes that might cause it, although it seems flakey and inconsistent.  I'd be in favour of dropping the token on login and providing brute force protection - which is far more likely to be a threat.

**Comment by stefaweb** [comment]
> If we reconnect to MW only after a long period, we got the message.

Hello! Just to say that the problem is repetitive.

**Comment by shawniverson** [comment]
Yeah, a good reason to do away with it and focus on brute force issues instead.

**Comment by endelwar** [comment]
I was ~~spying~~ taking inspiration from major frameworks (not because they are big, but because they have a lot of good and bad eyes on them) and all of them keep token on login page: if it expires the user is redirected on login page again, token regenerated and user warned about.
Another viable way is to use autorefresh on login page, so that token is always fresh: the autorefresh code should be suspended on user interaction (i.e. typing her username and password) and countdown restarted from initial value

**Comment by endelwar** [comment]
Also a bruteforce protection is a good idea, but needs a lot of work to be done: is it worth on 1.2 ?

**Comment by stefaweb** [comment]
Users need 1.2.1 asap...

**Comment by stefaweb** [comment]
Personally on the dev server, the last develop is fine. The last boring problem was the looping with disconnect.
I'm just bored of not having the no expiry for status and queue with 1.2.1.
